### PR TITLE
A simpler impl for `ByteArrayEqualityComparer` and remove redundant impl

### PR DESCRIPTION
**Pull request by nan01ab**
A better and simpler impl for ByteArrayEqualityComparer:

1. Use `SeqenceEqual` for better performance and  simpler implementation;
Perf test on 4KiB equal bytes:
```
BenchmarkDotNet v0.13.12, macOS Ventura 13.7 (22H123) [Darwin 22.6.0]
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET SDK 8.0.203
  [Host]     : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX2


| Method                            | Mean     | Error   | StdDev  |
|---------------------------------- |---------:|--------:|--------:|
| TestByteArrayEquals_SequenceEqual | 102.2 ns | 2.03 ns | 2.26 ns |
| TestByteArrayEquals_Unsafe        | 268.5 ns | 3.41 ns | 3.02 ns |
```

2. Remove duplicated implementations;

## Type of change

- [x] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo.Extensions/ByteArrayEqualityComparer.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        public static readonly ByteArrayEqualityComparer Default = new();
+
+        public bool Equals(byte[]? x, byte[]? y)
+        {
+            if (ReferenceEquals(x, y)) return true;
+            if (x is null || y is null || x.Length != y.Length) return false;
+
+            return x.SequenceEqual(y);
+        }
```
Comment (shargon) [inline src/Neo.Extensions/ByteArrayEqualityComparer.cs:28]: Linq is always slower

Comment (nan01ab) [reply src/Neo.Extensions/ByteArrayEqualityComparer.cs:28]: > Linq is always slower, seems weird to me

`SequenceEqual ` can use SIMD, so it faster.

File: src/Neo.Extensions/ByteArrayEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        public int GetHashCode(byte[] obj)
+        {
+            return (int)XxHash3.HashToUInt64(obj);
```
Comment (shargon) [inline src/Neo.Extensions/ByteArrayEqualityComparer.cs:32]: We should wait for https://github.com/neo-project/neo/pull/3531 and reuse the method

**Review by shargon — APPROVED** [review]

**Review by Jim8y — DISMISSED** [review]

**Comment by shargon** [comment]
Wait for https://github.com/neo-project/neo/pull/3503

**Review by shargon — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

**Pull request merged**